proposals/0072-feature-gate-threshold-automation.md

```diff
@@ -0,0 +1,193 @@
+In the future, this authority could be replaced by validator governance.
+
+The multi-signature authority stages a feature for activation by invoking the
+Feature Gate program's `StageFeatureForActivation` instruction. This instruction
+expects:
+
+- Data: The feature ID
+- Accounts:
+  - Staged Features PDA: writable
+  - Multi-signature authority: signer 
+
+One single PDA will be used to store two lists of features. Its data will be
+structured as follows:
+
+- 8 bytes for the current epoch.
+- A fixed-size list of size 8 containing 32-byte feature IDs staged for
+  activation at the end of the *current epoch*.
+- A fixed-size list of size 8 containing the `u64` value of total stake support
+  for each feature ID in the prior list.
+- 8 bytes for the next epoch.
+- A fixed-size list of size 8 containing 32-byte feature IDs staged for
+  activation at the end of the *next epoch*.
+
+`StageFeatureForActivation` will add the provided feature ID to the **next
+epoch's** set of staged features.
+
+An example of how this account's state might work to coordinate both lists can
+be found here:
+
+<https://github.com/buffalojoec/feature-gate-data-poc/blob/master/src/lib.rs>
+
+The Staged Features PDA could be derived simply from one literal seed:
+`"staged_features"`.
+
+### Step 3: Signaling Support for Staged Features
+
+With an on-chain reference point to determine the features staged for activation
+for a particular epoch, nodes can signal their support for the staged features
+supported by their software.
+
+A node signals its support for staged features by invoking the Feature Gate
+program's `SignalSupportForStagedFeatures` instruction. This instruction
+expects:
+
+- Data: A `u8` bit mask of the staged features.
+- Accounts:
+  - Staged Features PDA: writable
+  - Vote account: signer
+
+A `1` bit represents support for a feature. For example, for staged features
+`[A, B, C, D, E, F, G, H]`, if a node wishes to signal support for all features
+except `E` and `H`, their `u8` value would be 246, or `11110110`.
+
+A node's submitted bit mask is then used to add that node's current epoch stake
+to the Staged Features PDA for each feature for which it has signaled support.
+This is done by using the Get Epoch Stake Syscall.
+
+Nodes should send a transaction containing this instruction at some arbitrary
+point during the epoch at least 128 slots before the end of the epoch and on
+startup after any reboot.
+
+Note: If a feature is revoked, the list of staged features will not change, and
+nodes may still signal support for this feature. However, the runtime will not
+activate this feature if its corresponding feature account no longer exists
+on-chain.
+
+### Step 4: Feature Activation
+
+During the epoch rollover, the runtime can simply load the Staged Features PDA
+and calculate the stake - as a percentage of the total epoch stake - in support
+for each feature ID to determine which staged features to activate.
```

Comment:
Does the runtime also need to check which features are revoked, and skip activating these? How would it do that (where would it look)?

Reply:
They wouldn't exist on-chain, so they're not in the Bank's feature set! Therefore something like `new_feature_activations.contains(..)` would resolve to `false`. This being an Agave-specific implementation bit, of course.

If you think we need to be more specific on this for specification purposes, I'm happy to add more context here.

Reply:
I do mention this earlier in the document, here:

https://github.com/solana-foundation/solana-improvement-documents/pull/72/files#diff-b798a8c9dca99c07b780074b5abdc2c8d81722af0a57dcdde447fbe96bbc8ad1R143-R146

But if you think this should move somewhere else I'm okay with that.

Reply:
I think re-iterating that the runtime will not activate revoked features in the Step 4 section would be great, thanks!